### PR TITLE
Enable database read splitting for high volume controller actions.

### PR DESCRIPTION
**Pull request by sureshc**
CPU Utilization on the primary (writer) production database instance [currently averages about 30% ](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2:graph=~(metrics~(~(~'AWS*2fRDS~'CPUUtilization~'Role~'WRITER~'DBClusterIdentifier~'production-aurora-cluster~(stat~'Average))~(~'...~'READER~'.~'.~(visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~title~'CPU~yAxis~(left~(min~0~max~100)~right~(min~0))~period~60~annotations~(horizontal~(~(label~'Alarm~value~40)))~start~'2022-01-21T13*3a00*3a00.000Z~end~'2022-01-21T21*3a00*3a00.000Z~stat~'Maximum))during peak usage time periods. Historically, we have considered database CPU utilization above 40% to be an abnormally high level that should trigger paging the on call engineer.

image

Consider offloading read queries from the writer instance to the read replica instance by enabling read splitting on additional controller actions. Out of this [list of High Throughput](https://onenr.io/0NgR7rXBXwo) dashboard endpoints, the following are high throughput Rails controller actions that do not yet have read splitting enabled:
* levels#get_rubric
* api/v1/teacher_feedbacks#get_feedbacks
* scripts#show

## Testing story

We don't have database read splitting enabled on any environment other than production. Perhaps these changes are low risk because these controller actions are mostly read only?

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/levels_controller.rb [hunk 1]
```diff
@@ -59,6 +59,8 @@ class LevelsController < ApplicationController
     Weblab
   ]
 
+  use_reader_connection_for_route(:get_rubric)
```
Comment (Hamms) [inline dashboard/app/controllers/levels_controller.rb]: nit: I'd prefer for this line to be adjacent to the method definition

**Review by Hamms — APPROVED** [review]
Something else we might consider is directing the various model `summarize` methods to the read replica; I don't think it's a coincidence that most of our high-throughput controller routes end up calling one of them.

That being said, this LGTM!

File: dashboard/app/controllers/levels_controller.rb (hunk 1 above)
Comment (sureshc) [reply dashboard/app/controllers/levels_controller.rb]: Oops. Yes. I fixed this mistake in the teacher feedbacks controller, but forgot to apply that to this one.

**Review by jamescodeorg — APPROVED** [review]
LGTM, thanks for fixing!

(@maureensturgeon, we may also want to take a look at reducing the cost of the feedback queries at some point)

One more idea -- Suresh, do you think it would worthwhile to surface the most expensive queries data more regularly (e.g. in a weekly or bi-monthly Slack message)?

**Comment by sureshc** [comment]
> One more idea -- Suresh, do you think it would worthwhile to surface the most expensive queries data more regularly (e.g. in a weekly or bi-monthly Slack message)?

@jamescodeorg - Yes. Infrastructure should be able to find a way to provide reporting to all feature teams about which endpoints receive the most HTTP requests, and which database queries have the highest throughput or impact.

**Comment by sureshc** [comment]
This change reduced SELECT throughput on the writer by 1-2K/minute.
1057DC6E-BFAB-4111-92F3-5515E02A550C

**Comment by sureshc** [comment]
This decreased average CPU utilization on the writer by a few percentage points.
416F16D1-2215-473B-97C4-A8975282A6A4

**Comment by sureshc** [comment]
No significant change in latency for the controller actions that were modified.
F085B5EE-2EA8-4CDD-AEEF-5BAE38249330
30517491-FE94-443C-8A04-D7E5A11B33F9
45D31E90-5BEC-4428-B4C5-C32A0B386394

**Pull request merged**